### PR TITLE
drivaerml: Lion optimizer sweep at 4L/512d

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The DrivAerML baseline (3.997% val surface_rel_l2_pct, PR #2898) uses AdamW. On TandemFoil, switching to Lion with gc=1.0 and WD=1e-2 was the decisive improvement. Lion has never been tested on DrivAerML. Given the small effective batch size (batch-size=1, max-train-batches=394) and long training horizon (SENPAI_MAX_EPOCHS=9999), Lion's sign-based update rule may provide more stable gradient signal than AdamW — especially for the 3D surface pressure distribution where gradients can be noisy.

## Current Baseline

**DrivAerML: 3.997% val surface_rel_l2_pct (ep467, PR #2898)**

Baseline command:
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py   --dataset drivaerml   --optimizer adamw   --lr 5e-4   --cosine-t-max 30   --no-use-ema   --enable-fourier   --model-layers 4   --model-hidden-dim 512   --model-heads 8   --epochs 999   --batch-size 1   --drivaerml-train-surface-points 50000   --drivaerml-eval-surface-points 50000   --max-train-batches 394   --max-eval-batches 200
```

Primary metric: `val_primary/surface_rel_l2_pct` (lower is better). Target: beat 3.997%.

## Experiment Plan

Fix: 4L/512d/8H, no-EMA, Fourier, batch-size=1, all DrivAerML data flags. Sweep Lion lr and also include a gc+WD variant:

| Run | Optimizer | lr | cosine-t-max | gc | WD |
|---|---|---|---|---|---|
| A | Lion | 1e-4 | 30 | 1.0 | 1e-2 |
| B | Lion | 2e-4 | 30 | 1.0 | 1e-2 |
| C | Lion | 3e-4 | 30 | 1.0 | 1e-2 |
| D | Lion | 5e-4 | 30 | 1.0 | 1e-2 |
| E | Lion | 1e-4 | 50 | 1.0 | 1e-2 |
| F | Lion | 2e-4 | 50 | 1.0 | 1e-2 |
| G | AdamW | 5e-4 | 30 | 1.0 | 1e-2 | (reference — adds gc+WD to baseline) |
| H | AdamW | 3e-4 | 30 | 1.0 | 1e-2 | (reference — lower lr + gc+WD) |

## Instructions

**Run A (Lion lr=1e-4, T_max=30, gc=1.0, WD=1e-2):**
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py   --dataset drivaerml   --optimizer lion   --lr 1e-4   --cosine-t-max 30   --grad-clip 1.0   --weight-decay 1e-2   --no-use-ema   --enable-fourier   --model-layers 4   --model-hidden-dim 512   --model-heads 8   --epochs 999   --batch-size 1   --drivaerml-train-surface-points 50000   --drivaerml-eval-surface-points 50000   --max-train-batches 394   --max-eval-batches 200   --wandb_group drivaerml_lion_sweep
```

**Run B:** Same as A but `--lr 2e-4`

**Run C:** Same as A but `--lr 3e-4`

**Run D:** Same as A but `--lr 5e-4`

**Run E:** Same as A but `--cosine-t-max 50`

**Run F:** Same as B but `--cosine-t-max 50`

**Run G (AdamW reference with gc+WD):**
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py   --dataset drivaerml   --optimizer adamw   --lr 5e-4   --cosine-t-max 30   --grad-clip 1.0   --weight-decay 1e-2   --no-use-ema   --enable-fourier   --model-layers 4   --model-hidden-dim 512   --model-heads 8   --epochs 999   --batch-size 1   --drivaerml-train-surface-points 50000   --drivaerml-eval-surface-points 50000   --max-train-batches 394   --max-eval-batches 200   --wandb_group drivaerml_lion_sweep
```

**Run H:** Same as G but `--lr 3e-4`

## What to Report

For each run, report:
- `val_primary/surface_rel_l2_pct` at best validation epoch
- `test_primary/surface_rel_l2_pct` from best-val checkpoint
- WandB run URL

Key questions: Does Lion beat AdamW on DrivAerML? Does gc=1.0 + WD=1e-2 help AdamW on DrivAerML (currently not applied in baseline)?

## Suggested Follow-ups

If Lion wins: sweep WD (1e-3, 1e-2, 1e-1) and try 5L/512d. If AdamW+gc+WD wins: that's a baseline improvement.

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT piccolo:

## Results

### Summary

8-run Lion optimizer sweep on DrivAerML at 4L/512d/8H. All runs used gc=1.0, WD=1e-2, cosine scheduler, no-EMA, Fourier features, batch-size=1, 50k surface points, max-train-batches=394, max-eval-batches=200.

**No run beat the 3.997% baseline.** Best was Run A (Lion lr=1e-4, T_max=30) at **5.129%** val surface_rel_l2_pct.

### Results Table

| Run | Optimizer | lr | T_max | Best Val % | Best Epoch | Test % | Epochs | Status |
|-----|-----------|------|-------|-----------|------------|--------|--------|--------|
| **A** | **Lion** | **1e-4** | **30** | **5.129** | **144** | **6.360** | **147** | **Stable** |
| B | Lion | 2e-4 | 30 | 8.979 | 34 | 370.663 | 146 | Diverged ~ep69 |
| C | Lion | 3e-4 | 30 | 24.401 | 20 | 168.601 | 145 | Never stable |
| D | Lion | 5e-4 | 30 | 37.732 | ~7 | N/A | 20 | Diverged immediately |
| E | Lion | 1e-4 | 50 | 6.386 | 105 | 68.047 | 142 | Diverged ~ep120 |
| F | Lion | 2e-4 | 50 | 7.855 | 89 | 146.076 | 144 | Diverged ~ep100 |
| G | AdamW | 5e-4 | 30 | 6.844 | 130 | 9.273 | 145 | Diverged ep115, partial recovery |
| **H** | **AdamW** | **3e-4** | **30** | **6.242** | **139** | **7.748** | **147** | **Stable** |

**Baseline: 3.997% val surface_rel_l2_pct at epoch 467 (AdamW lr=5e-4, no gc/WD, PR #2898)**

### Commands Used

**Run A (representative):**
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=999 SENPAI_TIMEOUT_MINUTES=360 CUDA_VISIBLE_DEVICES=0 python train.py \
  --dataset drivaerml --agent piccolo \
  --wandb-name "piccolo/lion-1e4-tmax30-A" --wandb-group "drivaerml_lion_sweep" \
  --optimizer lion --lr 1e-4 --cosine-t-max 30 --grad-clip 1.0 --weight-decay 1e-2 \
  --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200
```
Runs B-F: same as A with varied `--lr` and `--cosine-t-max` per the experiment plan.
Runs G-H: same but `--optimizer adamw` with respective `--lr`.

### Peak Memory

~5.8 GB per GPU (Lion runs ~5.6 GB, AdamW runs ~5.8 GB due to additional moment buffer).

### W&B Run IDs

| Run | W&B URL |
|-----|---------|
| A | https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/znghuaxq |
| B | https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/woyghctl |
| C | https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/y8cs088d |
| D | https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/75zvphzq |
| E | https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/bo7ml3xw |
| F | https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/0eak7iw9 |
| G | https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/iwpa823q |
| H | https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/ox32oa4t |

W&B group: `drivaerml_lion_sweep`

### What Happened

**Lion is viable on DrivAerML but did not beat the baseline.** Run A (Lion lr=1e-4, T_max=30) achieved 5.129% at epoch 144 — 28% above the 3.997% baseline. However, the baseline had 467 epochs vs Run A's 147, so the gap may partly reflect insufficient training time.

**Key findings:**

1. **Lion is extremely LR-sensitive on DrivAerML.** Only lr=1e-4 was stable long-term. lr=2e-4 diverged at epochs 69-100, lr=3e-4 never stabilized, and lr=5e-4 diverged within 7 epochs. This is a much narrower stable range than AdamW, likely because Lion's sign-based updates create large effective step sizes when combined with the noisy 3D surface pressure gradients.

2. **T_max=30 >> T_max=50 for stability.** Even at the only stable lr=1e-4, increasing T_max from 30→50 (Run E) caused divergence at ~epoch 120. With 394 batches/epoch, T_max=30 gives ~13 cosine cycles per epoch vs ~8 with T_max=50 — the more frequent LR resets appear essential for Lion's stability on this dataset.

3. **Adding gc+WD to AdamW is destabilizing at lr=5e-4.** Run G (AdamW lr=5e-4 + gc=1.0 + WD=1e-2) diverged at epoch 115 — the baseline uses the same lr without gc/WD and runs stably for 467+ epochs. The weight decay penalty likely pushes AdamW's effective updates too large at this lr. At lr=3e-4 (Run H), AdamW + gc+WD was stable but only reached 6.242%, much worse than the 3.997% baseline at lr=5e-4 without gc/WD.

4. **Run A's learning curve was still improving at timeout.** Best val went from 5.49% at epoch 132 to 5.13% at epoch 144. If given the full 467+ epochs, it could potentially approach the baseline. But the LR sensitivity suggests Lion has less headroom for DrivAerML than AdamW.

### Suggested Follow-ups

1. **Run A for longer** (SENPAI_MAX_EPOCHS=9999, SENPAI_TIMEOUT_MINUTES=720) to see if Lion lr=1e-4 can actually match or beat 3.997% with enough epochs.
2. **Try Lion lr=5e-5 and lr=7e-5** — given the extreme LR sensitivity, the optimum may actually be below 1e-4.
3. **Try Lion without WD** (WD=0) — the baseline AdamW doesn't use WD. Lion's decoupled WD may be hurting rather than helping.
4. **Do NOT add gc+WD to the AdamW baseline** — Run G showed this is destabilizing at lr=5e-4.

**Comment by tcapelle** [comment]
ADVISOR: Closing as a dead end. The Lion optimizer sweep across 8 LR/T_max combinations failed to beat the current DrivAerML baseline of 3.997%. Best result was Run A (Lion lr=1e-4, T_max=30) at 5.129% — 28% above baseline. The other runs either diverged (lr=2e-4, lr=3e-4, lr=5e-4) or stalled at high loss.

Key finding that must be documented for future work: adding gc=1.0 + wd=1e-2 to the AdamW baseline at lr=5e-4 (Run G) caused divergence at ep115. The current DrivAerML baseline uses AdamW with no gradient clipping and no weight decay at lr=5e-4 — this must not be changed without carefully reducing the LR first.

Lion is viable on DrivAerML only at very low LR (1e-4) but is too slow to converge and too LR-sensitive to compete with the current AdamW baseline within the training budget. There is no path forward from this direction without a fundamentally different setup. GPU time is better spent elsewhere. Thank you for the thorough sweep.
